src/backend/distributed/transaction/worker_transaction.c

```diff
@@ -0,0 +1,479 @@
+	{
+		/*
+		 * Any failure here will cause local changes to be rolled back,
+		 * and may leave a prepared transaction on the remote node.
+		 */
+
+		PrepareRemoteTransactions(workerConnectionList);
+
+		/*
+		 * We are now ready to commit the local transaction, followed
+		 * by the remote transaction. As a final step, write commit
+		 * records to a table. If there is a last-minute crash
+		 * on the local machine, then the absence of these records
+		 * will indicate that the remote transactions should be rolled
+		 * back. Otherwise, the presence of these records indicates
+		 * that the remote transactions should be committed.
+		 */
+
+		LogPreparedTransactions(workerConnectionList);
+
+		return;
+	}
+	else if (event == XACT_EVENT_COMMIT)
+	{
+		/*
+		 * A failure here may cause some prepared transactions to be
+		 * left pending. However, the local change have already been
+		 * committed and a commit record exists to indicate that the
+		 * remote transaction should be committed as well.
+		 */
+
+		CommitRemoteTransactions(workerConnectionList, false);
+
+		/*
+		 * At this point, it is safe to remove the transaction records
+		 * for all commits that have succeeded. However, we are no
+		 * longer in a transaction and therefore cannot make changes
+		 * to the metadata.
+		 */
+	}
+	else if (event == XACT_EVENT_ABORT)
+	{
+		/*
+		 * A failure here may cause some prepared transactions to be
+		 * left pending. The local changes have already been rolled
+		 * back and the absence of a commit record indicates that
+		 * the remote transaction should be rolled back as well.
+		 */
+
+		AbortRemoteTransactions(workerConnectionList);
+	}
+	else if (event == XACT_EVENT_PREPARE || event == XACT_EVENT_PRE_PREPARE)
+	{
+		/*
+		 * If we allow a prepare we might not get to the commit handler
+		 * in this session. We could resolve that if we intercept
+		 * COMMIT/ABORT PREPARED commands. For now, we just error out.
+		 */
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot prepare a transaction that modified "
+							   "distributed tables")));
+	}
+	else
+	{
+		return;
+	}
+
+	CloseConnections(workerConnectionList);
+
+	/*
+	 * Memory allocated in workerConnectionList will be reclaimed when
+	 * TopTransactionContext is released.
+	 */
+
+	workerConnectionList = NIL;
+}
+
+
+/*
+ * OpenWorkerTransactions opens connections to all primary workers and sends
+ * BEGIN commands. The returned TransactionConnection's are allocated in the
+ * top transaction context, such that they can still be used in the commit
+ * handler. The connections are made as the extension owner, such that they
+ * have write access to the Citus metadata tables.
+ */
+static List *
+OpenWorkerTransactions(void)
+{
+	ListCell *workerNodeCell = NULL;
+	List *connectionList = NIL;
+	MemoryContext oldContext = NULL;
+	List *workerList = NIL;
+
+	/*
+	 * A cluster_activate might be in progress which will invalidate
```

Comment:
I guess this is a bit of a futuristic comment right now :). Maybe add a general description of "when we start syncing metadata to a new node"